### ISSUE TITLE
build functions for objects are not friendly with inherited classes

### ISSUE THREAD

**Issue by kbenoit** [quanteda/quanteda#1878]
**sentometrics** is breaking in the current `master` because our build functions drop additional class attributes. In 
https://github.com/sborms/sentometrics/blob/master/R/sentocorpus.R#L141-L142
for instance, `sento_corpus()` creates a quanteda corpus and then adds an additional class label (after giving it some docvars that are special for a sento_corpus).

But as you can see below, basic corpus functions that **sentocorpus** uses, such as `corpus_sample()`, erase the added class label. 

More generally, this means that additional package developers cannot create child classes (inasmuch as such a notion exists in the primitive S3 system) of **quanteda** objects without having our functions erase their added attributes. If the objects on which we build **quanteda**'s objects behaved like this, we could never have built our package. This seems like something we should strive to avoid generally.

``` r
library("quanteda")
## Package version: 2.0.0

corp <- corpus(letters[1:3],
  docvars = data.frame(dv = 11:13),
  meta = list(mymeta = "Test meta")
)

mycorp <- corp
class(mycorp) <- c("myclass", class(mycorp))

class(corpus_sample(mycorp, size = 2))
## [1] "corpus"
class(corpus_subset(mycorp, dv == 1))
## [1] "corpus"

docvars(mycorp, "dv2") <- c("one", "two", "three")
class(mycorp)
## [1] "myclass" "corpus"
```

**Comment by kbenoit** [comment]
Our S4 objects suffer the same problem. (Here, it happens in the `build_dfm()` call in `subset_dfm()`.)

``` r
dfmat <- dfm(corp)

setClass("dfmchild",
  slots = c(
    myslot1 = "character",
    myslot2 = "integer"
  ),
  contains = c("dfm")
)

dfmatchild <- as(dfmat, "dfmchild")
dfmatchild@myslot1 <- "test"
dfmatchild@myslot2 <- "42"
## Error in (function (cl, name, valueClass) : assignment of an object of class "character" is not valid for @'myslot2' in an object of class "dfmchild"; is(value, "integer") is not TRUE

dfm_sample(dfmatchild, size = 2) %>%
  is("dfmchild")
## [1] FALSE
dfm_sample(dfmatchild, size = 2) %>%
  slotNames()
## [1] "docvars"  "meta"     "i"        "p"        "Dim"      "Dimnames" "x"       
## [8] "factors"
```

**Comment by kbenoit** [comment]
A solution could be to call `build_x()` only when the constructor is called, and `rebuild_x()` when it is modified.

**Comment by kbenoit** [comment]
We solved the S3 issues in 500eb08 but could do better with the S4 objects.
